### PR TITLE
Add API to deprecate local unit

### PR DESCRIPTION
**Pull request by Rup-Narayan-Rajbanshi**
Addresses
  - https://github.com/IFRCGo/go-api/issues/2338

## Changes
  - Add api for depricate local unit and revert depricate.

## Checklist
Things that should succeed before merging.

- [x] Updated/ran unit tests
- [ ] Updated CHANGELOG.md

## Release

If there is a version update, make sure to tag the repository with the latest version.

### CODE REVIEW

**Review by szabozoltan69 — COMMENTED** [review]
Please replace the string "epricate" to "eprecate" where it occurs in the files. :-)

File: local_units/views.py [hunk 1]
```diff
@@ -214,6 +215,32 @@ def get_latest_changes(self, request, pk=None, version=None):
         serializer = LocalUnitChangeRequestSerializer(change_request, context={"request": request})
         return response.Response(serializer.data)
 
+    @extend_schema(request=LocalUnitDeprecateSerializer)
+    @action(detail=True, methods=["put"], url_path="deprecate")
```
Comment (susilnem) [inline local_units/views.py:219]:
```suggestion
    @action(detail=True, methods=["post"], url_path="deprecate")
```

File: local_units/views.py [hunk 2]
```diff
@@ -214,6 +215,32 @@ def get_latest_changes(self, request, pk=None, version=None):
...
+        if serializer.is_valid():
+            serializer.save(instance)
+            return response.Response(
+                {"message": f"Object {instance.id}-{instance.is_deprecated} deprecated successfully."},
+                status=status.HTTP_200_OK,
+            )
+        return response.Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+    @extend_schema(responses=PrivateLocalUnitSerializer)
+    @action(detail=True, methods=["put"], url_path="revert-deprecate")
```
Comment (susilnem) [inline local_units/views.py:233]:
```suggestion
    @action(detail=True, methods=["post"], url_path="revert-deprecate")
```

File: local_units/views.py [hunk 3]
```diff
@@ -214,6 +215,32 @@ def get_latest_changes(self, request, pk=None, version=None):
...
+    @extend_schema(request=LocalUnitDeprecateSerializer)
+    @action(detail=True, methods=["put"], url_path="deprecate")
+    def deprecate(self, request, pk=None):
+        """Deprecate local unit object object"""
+        instance = self.get_object()
+        serializer = LocalUnitDeprecateSerializer(data=request.data)
+        if serializer.is_valid():
+            serializer.save(instance)
+            return response.Response(
+                {"message": f"Object {instance.id}-{instance.is_deprecated} deprecated successfully."},
```
Comment (susilnem) [inline local_units/views.py:227]:
```suggestion
                {"message": f"Object {instance.id} is deprecated successfully."},
```

File: local_units/views.py [hunk 4]
```diff
@@ -214,6 +215,32 @@ def get_latest_changes(self, request, pk=None, version=None):
...
+                status=status.HTTP_200_OK,
+            )
+        return response.Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
```
Comment (susilnem) [inline local_units/views.py:230]:
```suggestion
        serializer.is_valid(raise_exception=True)
        return response.Response(serializer.data, status=status.HTTP_200_OK)
```

**Review by susilnem — CHANGES_REQUESTED** [review]
Minor changes

**Review by susilnem — APPROVED** [review]
LGTM!

File: local_units/serializers.py [hunk 5]
```diff
@@ -550,3 +550,16 @@ class Meta:
             "created_by_details",
             "previous_data",
         )
+
+
+class LocalUnitDeprecateSerializer(serializers.Serializer):
```
Comment (thenav56) [inline local_units/serializers.py:555]: Let's use ModelSerializer with fields mentioned below

File: local_units/views.py [hunk 6]
```diff
@@ -214,6 +215,42 @@ def get_latest_changes(self, request, pk=None, version=None):
...
+    @extend_schema(request=LocalUnitDeprecateSerializer, responses=None)
+    @action(
+        detail=True,
+        methods=["post"],
+        url_path="deprecate",
+        serializer_class=LocalUnitDeprecateSerializer,
+        permission_classes=[permissions.IsAuthenticated, DenyGuestUserPermission],
+    )
+    def deprecate(self, request, pk=None):
+        """Deprecate local unit object object"""
```
Comment (thenav56) [inline local_units/views.py:227]: Error if already deprecate

**Review by thenav56 — CHANGES_REQUESTED** [review]

File: local_units/serializers.py [hunk 7]
```diff
@@ -550,3 +550,29 @@ class Meta:
...
+class LocalUnitDeprecateSerializer(serializers.ModelSerializer):
+
+    class Meta:
+        model = LocalUnit
+        fields = ("deprecated_reason", "deprecated_reason_overview")
+
+    def validate(self, attrs):
+        instance = self.instance
+        if instance and instance.is_deprecated:
+            raise serializers.ValidationError("This object is already deprecated.")
```
Comment (susilnem) [inline local_units/serializers.py:564]: Confirm error message with FE

File: local_units/serializers.py [hunk 8]
```diff
@@ -550,3 +550,29 @@ class Meta:
...
+
+        if not attrs.get("deprecated_reason"):
+            raise serializers.ValidationError("A reason must be provided when deprecating an object.")
```
Comment (susilnem) [inline local_units/serializers.py:567]: If required, just use serializer field

File: local_units/serializers.py [hunk 9]
```diff
@@ -550,3 +550,29 @@ class Meta:
...
+
+        return attrs
```
Comment (susilnem) [inline local_units/serializers.py:569]: raise validation error for create method.

File: local_units/serializers.py [hunk 10]
```diff
@@ -550,3 +550,29 @@ class Meta:
...
+
+    def update(self, instance, validated_data):
+        instance.is_deprecated = True
+        instance.deprecated_reason = validated_data.get("deprecated_reason", instance.deprecated_reason)
+        instance.deprecated_reason_overview = validated_data.get(
+            "deprecated_reason_overview", instance.deprecated_reason_overview
+        )
+        instance.save()
```
Comment (susilnem) [inline local_units/serializers.py:577]: use update fields

**Review by susilnem — CHANGES_REQUESTED** [review]

**Review by susilnem — CHANGES_REQUESTED** [review]
Minor changes

**Pull request merged**